flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetIntersect.scala

```diff
@@ -83,66 +82,68 @@ class DataSetIntersect(
       tableEnv: BatchTableEnvironment,
       expectedType: Option[TypeInformation[Any]]): DataSet[Any] = {
 
-    var leftDataSet: DataSet[Any] = null
-    var rightDataSet: DataSet[Any] = null
+    val leftDataSet: DataSet[Any] = left.asInstanceOf[DataSetRel].translateToPlan(tableEnv)
+    val rightDataSet: DataSet[Any] = right.asInstanceOf[DataSetRel].translateToPlan(tableEnv)
 
-    expectedType match {
-      case None =>
-        leftDataSet = left.asInstanceOf[DataSetRel].translateToPlan(tableEnv)
-        rightDataSet =
-          right.asInstanceOf[DataSetRel].translateToPlan(tableEnv, Some(leftDataSet.getType))
-      case _ =>
-        leftDataSet = left.asInstanceOf[DataSetRel].translateToPlan(tableEnv, expectedType)
-        rightDataSet = right.asInstanceOf[DataSetRel].translateToPlan(tableEnv, expectedType)
-    }
-
-    val config = tableEnv.getConfig
-
-    val returnType = determineReturnType(
-      getRowType,
-      expectedType,
-      config.getNullCheck,
-      config.getEfficientTypeUsage)
-
-    val generator = new CodeGenerator(
-      config,
-      false,
-      leftDataSet.getType,
-      Some(rightDataSet.getType))
-
-    val conversion = generator.generateConverterResultExpression(
-      returnType,
-      left.getRowType.getFieldNames)
+    val coGroupedDs = leftDataSet.coGroup(rightDataSet)
 
+    val leftType = leftDataSet.getType
+    val rightType = rightDataSet.getType
 
-    val body = s"""
-            |${conversion.code}
-            |${generator.collectorTerm}.collect(${conversion.resultTerm});
-            |""".stripMargin
+    // If it is atomic type, the field expression need to be "*".
```

Comment:
I think you can use `*` for all cases. Composite types should be (recursively) expanded to all atomic member types.